pkg/request/pool.go

```diff
@@ -216,39 +234,231 @@ func (rp *Pool) NextRequests(ctx context.Context) [][]byte {
 func (rp *Pool) Close() {
+	rp.lock.Lock()
+	defer rp.lock.Unlock()
 	atomic.StoreUint32(&rp.closed, 1)
-	rp.Clear()
+	if rp.pending != nil {
+		rp.pending.Close()
+	}
+	rp.pending = nil
+	rp.batchStore = nil
 }
 
-// StopTimers stops all the timeout timers attached to the pending requests, and marks the pool as "stopped".
-// This which prevents submission of new requests, and renewal of timeouts by timer go-routines that where running
-// at the time of the call to StopTimers().
-func (rp *Pool) StopTimers() {
+func (rp *Pool) isClosed() bool {
+	return atomic.LoadUint32(&rp.closed) == 1
+}
+
+// Halt stops the callbacks of the first and second strikes.
+func (rp *Pool) Halt() {
 	atomic.StoreUint32(&rp.stopped, 1)
+	if !rp.isBatchingEnabled() {
+		rp.pending.Stop()
+	}
+}
 
-	rp.logger.Debugf("Stopped all timers")
+func (rp *Pool) isStopped() bool {
+	return atomic.LoadUint32(&rp.stopped) == 1
 }
 
-// RestartTimers restarts all the timeout timers attached to the pending requests, as RequestForwardTimeout, and re-allows
-// submission of new requests.
-func (rp *Pool) RestartTimers() {
+// Restart restarts the pool.
+// When batching is set to true the pool is expected to respond to NextRequests.
+func (rp *Pool) Restart(batching bool) {
 	rp.lock.Lock()
 	defer rp.lock.Unlock()
 
-	// TODO
-	rp.logger.Debugf("Restarted all timers")
+	defer atomic.StoreUint32(&rp.stopped, 0)
+
+	rp.Halt()
+
+	batchingWasEnabled := rp.isBatchingEnabled()
+
+	if batchingWasEnabled && batching {
+		// if batching was already enabled there is nothing to do
+		return
+	}
+
+	if !batchingWasEnabled && !batching {
+		// if batching was not enabled anyway just reset timestamps of the pending store
+		rp.pending.ResetTimestamps()
+		return
+	}
+
+	rp.setBatching(batching) // change the batching
+
+	if batchingWasEnabled { // batching was enabled and now it is not
+		rp.moveToPendingStore()
+		return
+	}
+
+	// batching was not enabled but now it is
+	rp.moveToBatchStore()
+	return
+
+}
+
+func (rp *Pool) setBatching(enabled bool) {
+	if enabled {
+		atomic.StoreUint32(&rp.batchingEnabled, 1)
+	} else {
+		atomic.StoreUint32(&rp.batchingEnabled, 0)
+	}
+}
+
+func (rp *Pool) isBatchingEnabled() bool {
+	return atomic.LoadUint32(&rp.batchingEnabled) == 1
+}
+
+func (rp *Pool) moveToPendingStore() {
+	requests := make([][]byte, 0, rp.options.MaxSize)
+	rp.batchStore.ForEach(func(_, v interface{}) {
+		requests = append(requests, v.(*requestItem).request)
+	})
+	rp.pending = createPendingStore(rp.logger, rp.inspector, rp.options)
+	rp.pending.Init()
+	for _, req := range requests {
+		ctx, cancel := context.WithTimeout(context.Background(), rp.options.SubmitTimeout)
+		if err := rp.pending.Submit(req, ctx); err != nil {
+			rp.logger.Errorf("Could not submit request into pending store; error: %s", err)
+			cancel()
+			return
+		}
+		cancel()
+	}
+	rp.pending.Start()
+	rp.batchStore = nil
+	rp.semaphore = semaphore.NewWeighted(int64(rp.options.MaxSize))
```

Comment:
this part is a bit tricky, i think.

Should we not make sure we release all goroutines pending on the semaphore?

Reply:
This semaphore is only acquired once before submitting to batch store
Maybe I could just change this to `TryAcquire` and then I don't have to worry of pending inserts?

Reply:
While this would work, it will also make clients unable to wait for space to be available in the request pool and they will fail fast as a result. 

Correct me if i'm wrong, but, anything beyond the semaphore acquiring is lock free, so eventually, progress is guaranteed for all pending insertions. This also means that the only thing that keeps the memory from blowing up is this semaphore. 
If we initialize it from scratch, we are essentially further increasing the potential expansion of the memory as we now have the old pending insertions and also MaxSize new insertions.

Can we not preserve the semaphore and not recreate it? I think in the original implementation we don't do that... right?

Reply:
> Can we not preserve the semaphore and not recreate it? I think in the original implementation we don't do that... right?

I didn't understand this last sentence 
I added the recreate part...
Anyway what do you suggest I should do?